v2/bigtable-cdc-to-hbase/src/main/java/com/google/cloud/teleport/v2/templates/transforms/ConvertChangeStream.java

```diff
@@ -0,0 +1,287 @@
+      if (mutation.getEntries().size() == 0) {
+        throw new Exception("Change stream entries list is empty.");
+      }
+
+      byte[] hbaseRowKey = mutation.getRowKey().toByteArray();
+      RowMutations hbaseMutations = new RowMutations(hbaseRowKey);
+      // Check and convert entries to hbase mutations
+      for (Entry entry : mutation.getEntries()) {
+        if (entry instanceof SetCell) {
+          hbaseMutations.add(convertSetCell(hbaseRowKey, entry));
+        } else if (entry instanceof DeleteCells) {
+          hbaseMutations.add(convertDeleteCells(hbaseRowKey, entry));
+        } else if (entry instanceof DeleteFamily) {
+          hbaseMutations.add(convertDeleteFamily(hbaseRowKey, entry));
+        } else {
+          // All change stream entry types should be supported.
+          throw new Exception("Change stream entry is not a supported type for conversion.");
+        }
+      }
+      return hbaseMutations;
+    }
+
+    private static Put convertSetCell(byte[] hbaseRowKey, Entry entry) {
+      SetCell setCell = (SetCell) entry;
+      // Convert timestamp to milliseconds
+      long ts = convertMicroToMilliseconds(setCell.getTimestamp());
+      Put put = new Put(hbaseRowKey, ts);
+      put.addColumn(
+          convertUtf8String(setCell.getFamilyName()),
+          setCell.getQualifier().toByteArray(),
+          setCell.getValue().toByteArray());
+
+      return put;
+    }
+
+    private static Delete convertDeleteCells(byte[] hbaseRowKey, Entry entry) {
+      DeleteCells deleteCells = (DeleteCells) entry;
+
+      // Convert timestamp to milliseconds
+      long ts = convertMicroToMilliseconds(deleteCells.getTimestampRange().getEnd());
+      // If RowMutation is created with an unbounded timestamp, the mutation is meant to delete
+      // all versions of a cell up to the point of operation.
+      // This behavior is approximated by casting the Hbase delete as deleting up to current point
+      // in time.
+      if (deleteCells.getTimestampRange() == TimestampRange.unbounded()) {
+        ts = Time.now();
+      }
+
+      Delete delete = new Delete(hbaseRowKey, ts);
+      // Delete all versions of this column, which corresponds to Bigtable delete behavior.
+      // TODO: it is possible for CBT to delete single versions of column.
+      //  differentiate between addColumn and AddColumns
+      delete.addColumns(
+          convertUtf8String(deleteCells.getFamilyName()), deleteCells.getQualifier().toByteArray());
+
+      return delete;
+    }
+
+    private static Delete convertDeleteFamily(byte[] hbaseRowKey, Entry entry) {
+      // Bigtable deletefamily does not have a timestamp and relies on transaction sequence to
+      // delete everything before it.
+      // Hbase deletes operate from timestamps only. Therefore, we approximate Bigtable
+      // deletefamily to Hbase deletefamily with timestamp now().
+      long now = Time.now();
```

Comment:
same as above. Use timestamp from CBT,  it should be the time when CBT wrote that entry in its replication/WAL logs.

Reply:
@jackdingilian Double check my thinking here. You can process stream after "days", right? We shouldn't use stream processing time (Now) anywhere, rather we should base it on the time mutation was written to CBT WAL.

Reply:
@jackdingilian I don't see any timestamp for DeleteFamily or its parent class. If the commit timestamp exists, then we can also generalize to using that timestamp for Delete requests with unbounded timestamps.

Reply:
@vermas2012 Yeah that's correct, you could process it as far back as the change stream retention period if you're catching up.

@georgecma you can use the `commitTimestamp` on the `ChangeStreamMutation` that contains the DeleteFamily

Reply:
Changed implementation to use commitTimestamp from ChangeStreamMutation